### PR TITLE
[5.7] Throw error with proper message when dispatching dynamic method calls

### PR DESCRIPTION
**Pull request by JosephSilber**
### Intro

Laravel has _a lot_ of dynamic method calls, such as in Eloquent's `Model` class:

https://github.com/laravel/framework/blob/a1c2f82c2d38842a0296070b99d997be6d985ae3/src/Illuminate/Database/Eloquent/Model.php#L1564-L1571

This is how calling `User::where(...)` works, because the call to the `where` method is dynamically dispatched to the query builder.

### Problem

If you call a method on the model that doesn't exists, for example `User::missingMethod()`, you get an error saying that the method doesn't exist _on the query builder_ class:

```
BadMethodCallException: Call to undefined method Illuminate\Database\Query\Builder::missingMethod()
```

This message is quite cryptic, since the user hasn't called _any_ method on the query builder. This is especially confusing to newcomers, who don't even know about the builder and/or dynamic dispatch.

## Solution

This PR adds [a new `ForwardsCalls` trait](https://github.com/JosephSilber/framework/commit/525f780709026979e08ba5cfe0da91a6853075d8) to the `Support` module, which is able to intelligently catch and parse these errors. It'll convert the error to a new error with the message you would expect:

```
BadMethodCallException: Call to undefined method App\User::missingMethod()
```

I hooked up many of the framework's dynamic calls to this trait. After we merge this, I'm sure there are more places throughout the framework that we can use this.

### CODE REVIEW

**Comment by garygreen** [comment]
I'm sure there was another PR that did this... I was looking forward to it because it's far easier to debug. Not sure why the original one didn't get merged, or maybe it did and I'm thinking of something else 😳

**Comment by garygreen** [comment]
Ahh I found it, similar PR: https://github.com/laravel/framework/pull/20196

That was reverted in https://github.com/laravel/framework/pull/21929

Does this PR have similar problems?

**Comment by JosephSilber** [comment]
@garygreen I didn't know this was previously attempted...

> Does this PR have similar problems?

No. The `ForwardsCall` trait [intelligently checks](https://github.com/JosephSilber/framework/blob/97733713699f4de69dd2bf46d8cfb11de7f34a30/src/Illuminate/Support/Traits/ForwardsCalls.php#L15-L23) if the exception should be exchanged.

There's even [an explicit test](https://github.com/JosephSilber/framework/blob/97733713699f4de69dd2bf46d8cfb11de7f34a30/tests/Support/ForwardsCallsTest.php#L33-L40) here for it.

**Comment by thecrypticace** [comment]
You should attach the previous exception to the BadMethodCallException (the `$previous` argument) so that stack traces are properly preserved. You'll lose the context around the originating exception if you don't.

Something like this:

```php
throw new BadMethodCallException(sprintf(
 'Call to undefined method %s::%s()', static::class, $method
), $e->getCode(), $e);
```

**Comment by JosephSilber** [comment]
@thecrypticace the whole point here is that you _don't_ need the context of the original exception. If the object that the method is being deferred to doesn't have this method, then we don't want to forward the call there to begin with.

In other words, if there were a way for us to check this before even trying it, we would throw this `BadMethodCallException` right away. Since we can't, we first try it. But the exception being cought is an internal implementation that the caller shouldn't be aware of.

**Comment by garygreen** [comment]
@JosephSilber I think Jordan is referring to the stack trace / log, seems it's lost if you don't pass it on. Maybe a problem when it writes to the logger it won't have the full / previous stack trace leading up to the exception? Haven't tested it so maybe my understanding is wrong.

**Comment by thecrypticace** [comment]
That is indeed what I was referring to. You could perhaps attach it to the BadMethodCallException's own previous exception by using `->getPrevious()`. This would achieve the effect of hiding the implementation detail of the originating `BadMethodCallException` while basically preserving the trace for error reporting and logging purposes.

Though, with this type of error I guess It is less likely to be an issue because really an exception would have to lead to the bad method call and if you're testing your app hopefully it would get caught.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -1340,9 +1341,7 @@ public static function __callStatic($method, $parameters)
         }
 
         if (! isset(static::$macros[$method])) {
-            throw new BadMethodCallException(sprintf(
-                'Method %s::%s does not exist.', static::class, $method
-            ));
+            $this->throwBadMethodCallException($method);
```
Comment (austinkregel) [inline src/Illuminate/Database/Eloquent/Builder.php]: I don't know if anyone else caught this, but you can't call `$this` from a static method...

**Comment by JosephSilber** [comment]
Oh, good catch. I'll fix it later.

**Comment by garygreen** [comment]
@JosephSilber can you also fix the regex to support underscores + numbers in function names?

**Pull request merged**